Unity 2018.1 and IL2CPP

API doesnot work with Unity 2018.1 and IL2CPP enabled.

```
NotSupportedException: To marshal a managed method, please add an attribute named 'MonoPInvokeCallback' to the method definition.
  at Steamworks.NativeMethods.ISteamClient_SetWarningMessageHook (System.IntPtr instancePtr, Steamworks.SteamAPIWarningMessageHook_t pFunction) [0x00000] in <00000000000000000000000000000000>:0 
  at Steamworks.SteamClient.SetWarningMessageHook (Steamworks.SteamAPIWarningMessageHook_t pFunction) [0x00000] in <00000000000000000000000000000000>:0 
  at SteamManager.OnEnable () [0x00000] in <00000000000000000000000000000000>:0 
  at UnityEngine.GameObject.AddComponent (System.Type componentType) [0x00000] in <00000000000000000000000000000000>:0 
  at UnityEngine.GameObject.AddComponent[T] () [0x00000] in <00000000000000000000000000000000>:0 
  at SteamManager.get_Initialized () [0x00000] in <00000000000000000000000000000000>:0 
  at UnityEngine.SetupCoroutine.InvokeMoveNext (System.Collections.IEnumerator enumerator, System.IntPtr returnValueAddress) [0x00000] in <00000000000000000000000000000000>:0 
UnityEngine.GameObject:AddComponent(Type)
UnityEngine.GameObject:AddComponent()
SteamManager:get_Initialized()
UnityEngine.SetupCoroutine:InvokeMoveNext(IEnumerator, IntPtr)
```

```
NotSupportedException: IL2CPP does not support marshaling delegates that point to generic methods.
  at System.Runtime.InteropServices.Marshal.StructureToPtr (System.Object structure, System.IntPtr ptr, System.Boolean fDeleteOld) [0x00000] in <00000000000000000000000000000000>:0 
  at Steamworks.Callback`1[T].BuildCCallbackBase () [0x00000] in <00000000000000000000000000000000>:0 
  at Steamworks.Callback`1[T]..ctor (Steamworks.Callback`1+DispatchDelegate[T] func, System.Boolean bGameServer) [0x00000] in <00000000000000000000000000000000>:0 
  at Steamworks.Callback`1[T].Create (Steamworks.Callback`1+DispatchDelegate[T] func) [0x00000] in <00000000000000000000000000000000>:0 
  at SteamSocialAPI+<Start>c__Iterator0.MoveNext () [0x00000] in <00000000000000000000000000000000>:0 
  at UnityEngine.SetupCoroutine.InvokeMoveNext (System.Collections.IEnumerator enumerator, System.IntPtr returnValueAddress) [0x00000] in <00000000000000000000000000000000>:0 
 
(Filename: currently not available on il2cpp Line: -1)
```

Not work for me. 800+ errors like that:
`
Assets/Plugins/Steamworks.NET/autogen/SteamCallbacks.cs(24,30): error CS0051: Inconsistent accessibility: parameter type `Steamworks.CCallbackBaseVTable.RunCBDel' is less accessible than method `Steamworks.SteamCallbackBase.BuildVTable(Steamworks.CCallbackBaseVTable.RunCBDel, Steamworks.CCallbackBaseVTable.RunCRDel, Steamworks.CCallbackBaseVTable.GetCallbackSizeBytesDel)'
`

--- a/Plugins/Steamworks.NET/CallbackDispatcher.cs
+++ b/Plugins/Steamworks.NET/CallbackDispatcher.cs
@@ -75,6 +75,17 @@ namespace Steamworks {
 				foreach (var iCallback in unusedICallbacks) {
 					m_registeredCallbacks.Remove(iCallback);
 				}
+
+				unusedICallbacks.Clear();
+
+				foreach (var pair in m_registeredGameServerCallbacks) {
+					if (pair.Value.Count == 0)
+						unusedICallbacks.Add(pair.Key);
+				}
+
+				foreach (var iCallback in unusedICallbacks) {
+					m_registeredGameServerCallbacks.Remove(iCallback);
+				}
 			}
 		}
 

--- a/Plugins/Steamworks.NET/CallbackDispatcher.cs
+++ b/Plugins/Steamworks.NET/CallbackDispatcher.cs
@@ -25,29 +25,8 @@
 	#error You need to define STEAMWORKS_WIN, or STEAMWORKS_LIN_OSX. Refer to the readme for more details.
 #endif
 
-// Unity 32bit Mono on Windows crashes with ThisCall/Cdecl for some reason, StdCall without the 'this' ptr is the only thing that works..?
-#if (UNITY_EDITOR_WIN && !UNITY_EDITOR_64) || (!UNITY_EDITOR && UNITY_STANDALONE_WIN && !UNITY_64)
-	#define STDCALL
-#elif STEAMWORKS_WIN
-	#define THISCALL
-#endif
-
-// Calling Conventions:
-// Unity x86 Windows        - StdCall (No this pointer)
-// Unity x86 Linux          - Cdecl
-// Unity x86 OSX            - Cdecl
-// Unity x64 Windows        - Cdecl
-// Unity x64 Linux          - Cdecl
-// Unity x64 OSX            - Cdecl
-// Microsoft x86 Windows    - ThisCall
-// Microsoft x64 Windows    - ThisCall
-// Mono x86 Linux           - Cdecl
-// Mono x86 OSX             - Cdecl
-// Mono x64 Linux           - Cdecl
-// Mono x64 OSX             - Cdecl
-// Mono on Windows is probably not supported.
-
 using System;
+using System.Collections.Generic;
 using System.Runtime.InteropServices;
 
 namespace Steamworks {
@@ -62,19 +41,139 @@ namespace Steamworks {
 			Console.WriteLine(e.Message);
 #endif
 		}
+
+		private static Dictionary<int, List<Callback>> m_registeredCallbacks = new Dictionary<int, List<Callback>>();
+		private static Dictionary<int, List<Callback>> m_registeredGameServerCallbacks = new Dictionary<int, List<Callback>>();
+		private static Dictionary<ulong, CallResult> m_registeredCallResults = new Dictionary<ulong, CallResult>();
+		private static List<ulong> m_dispatchedApiCalls = new List<ulong>();
+		private static object m_sync = new object();
+		private static CallbackMsg_t m_callbackMsg; // Preallocated
+		private static GCHandle m_pCallbackMsg;
+
+		public static bool IsInitialized { get; private set; }
+
+		internal static void Initialize() {
+			NativeMethods.SteamAPI_ManualDispatch_Init();
+			m_pCallbackMsg = GCHandle.Alloc(m_callbackMsg, GCHandleType.Pinned);
+			IsInitialized = true;
+		}
+
+		internal static void Shutdown() {
+			if (m_pCallbackMsg.IsAllocated)
+				m_pCallbackMsg.Free();
+			IsInitialized = false;
+		}
+
+		public static void Prune() {
+			List<int> unusedICallbacks = new List<int>();
+			lock (m_sync) {
+				foreach (var pair in m_registeredCallbacks) {
+					if (pair.Value.Count == 0)
+						unusedICallbacks.Add(pair.Key);
+				}
+
+				foreach (var iCallback in unusedICallbacks) {
+					m_registeredCallbacks.Remove(iCallback);
+				}
+			}
+		}
+
+		internal static void Register(Callback cb) {
+			int iCallback = CallbackIdentities.GetCallbackIdentity(cb.GetCallbackType());
+			var callbacksRegistry = cb.IsGameServer ? m_registeredGameServerCallbacks : m_registeredCallbacks;
+			lock (m_sync) {
+				List<Callback> callbacksList;
+				if (!callbacksRegistry.TryGetValue(iCallback, out callbacksList)) {
+					callbacksList = new List<Callback>();
+					callbacksRegistry.Add(iCallback, callbacksList);
+				}
+
+				callbacksList.Add(cb);
+			}
+		}
+
+		internal static void Register(SteamAPICall_t asyncCall, CallResult cr) {
+			lock (m_sync) {
+				// Dispatch goes to last CallResult the SteamAPICall_t is set on
+				m_registeredCallResults[(ulong)asyncCall] = cr;
+			}
+		}
+
+		internal static void Unregister(Callback cb) {
+			int iCallback = CallbackIdentities.GetCallbackIdentity(cb.GetCallbackType());
+			var callbacksRegistry = cb.IsGameServer ? m_registeredGameServerCallbacks : m_registeredCallbacks;
+			lock (m_sync) {
+				List<Callback> callbacksList;
+				if (callbacksRegistry.TryGetValue(iCallback, out var callbacksList)) {
+					callbacksList.Remove(cb);
+				}
+			}
+		}
+
+		internal static void Unregister(SteamAPICall_t asyncCall) {
+			lock (m_sync) {
+				m_registeredCallResults.Remove((ulong)asyncCall);
+			}
+		}
+
+		internal static void RunFrame(bool isGameServer) {
+			if (!IsInitialized) throw new InvalidOperationException("Callback dispatcher is not initialized.");
+
+			HSteamPipe hSteamPipe = (HSteamPipe)(isGameServer ? NativeMethods.SteamGameServer_GetHSteamPipe() : NativeMethods.SteamAPI_GetHSteamPipe());
+			NativeMethods.SteamAPI_ManualDispatch_RunFrame(hSteamPipe);
+			var callbacksRegistry = isGameServer ? m_registeredGameServerCallbacks : m_registeredCallbacks;
+			lock (m_sync) {
+				while (NativeMethods.SteamAPI_ManualDispatch_GetNextCallback(hSteamPipe, m_pCallbackMsg.AddrOfPinnedObject())) {
+					try {
+						// Check for dispatching API call results
+						if (m_callbackMsg.m_iCallback == SteamAPICallCompleted_t.k_iCallback) {
+							SteamAPICallCompleted_t callCompletedCb;
+							Marshal.PtrToStructure(m_callbackMsg.m_pubParam, callCompletedCb);
+							IntPtr pTmpCallResult = Marshal.AllocHGlobal(callCompletedCb.m_cubParam);
+							bool bFailed;
+							try {
+								if (NativeMethods.SteamAPI_ManualDispatch_GetAPICallResult(hSteamPipe, callCompletedCb.m_hAsyncCall, pTmpCallResult, callCompletedCb.m_cubParam, callCompletedCb.m_iCallback, out bFailed)) {
+									if (m_registeredCallResults.TryGetValue((ulong)callCompletedCb.m_hAsyncCall, out CallResult cr)) {
+										cr.OnRunCallResult(pTmpCallResult, bFailed, (ulong)callCompletedCb.m_hAsyncCall);
+									}
+								}
+							} finally {
+								Marshal.FreeHGlobal(pTmpCallResult);
+								m_dispatchedApiCalls.Add((ulong)callCompletedCb.m_hAsyncCall);
+							}
+						} else {
+							if (callbacksRegistry.TryGetValue(m_callbackMsg.m_iCallback, out var callbacks)) {
+								foreach (var callback in callbacks) {
+									callback.OnRunCallback(m_callbackMsg.m_pubParam);
+								}
+							}
+						}
+					} catch (Exception e) {
+						ExceptionHandler(e);
+					} finally {
+						NativeMethods.SteamAPI_ManualDispatch_FreeLastCallback(hSteamPipe);
+						foreach (var call in m_dispatchedApiCalls) {
+							m_registeredCallResults.Remove(call);
+						}
+						m_dispatchedApiCalls.Clear();
+					}
+				}
+			}
+		}
 	}
 
-	public sealed class Callback<T> : IDisposable {
-		private CCallbackBaseVTable m_CallbackBaseVTable;
-		private IntPtr m_pVTable = IntPtr.Zero;
-		private CCallbackBase m_CCallbackBase;
-		private GCHandle m_pCCallbackBase;
+	public abstract class Callback {
+		public abstract bool IsGameServer { get; }
+		internal abstract Type GetCallbackType();
+		internal abstract void OnRunCallback(IntPtr pvParam);
+	}
 
+	public sealed class Callback<T> : Callback, IDisposable {
 		public delegate void DispatchDelegate(T param);
 		private event DispatchDelegate m_Func;
 
 		private bool m_bGameServer;
-		private readonly int m_size = Marshal.SizeOf(typeof(T));
+		private bool m_bIsRegistered;
 
 		private bool m_bDisposed = false;
 
@@ -98,7 +197,6 @@ namespace Steamworks {
 
 		public Callback(DispatchDelegate func, bool bGameServer = false) {
 			m_bGameServer = bGameServer;
-			BuildCCallbackBase();
 			Register(func);
 		}
 
@@ -115,14 +213,6 @@ namespace Steamworks {
 
 			Unregister();
 
-			if (m_pVTable != IntPtr.Zero) {
-				Marshal.FreeHGlobal(m_pVTable);
-			}
-
-			if (m_pCCallbackBase.IsAllocated) {
-				m_pCCallbackBase.Free();
-			}
-
 			m_bDisposed = true;
 		}
 
@@ -132,34 +222,28 @@ namespace Steamworks {
 				throw new Exception("Callback function must not be null.");
 			}
 
-			if ((m_CCallbackBase.m_nCallbackFlags & CCallbackBase.k_ECallbackFlagsRegistered) == CCallbackBase.k_ECallbackFlagsRegistered) {
+			if (m_bIsRegistered) {
 				Unregister();
-			}
-
-			if (m_bGameServer) {
-				SetGameserverFlag();
 			}
 
 			m_Func = func;
 
-			// k_ECallbackFlagsRegistered is set by SteamAPI_RegisterCallback.
-			NativeMethods.SteamAPI_RegisterCallback(m_pCCallbackBase.AddrOfPinnedObject(), CallbackIdentities.GetCallbackIdentity(typeof(T)));
+			CallbackDispatcher.Register(this);
+			m_bIsRegistered = true;
 		}
 
 		public void Unregister() {
-			// k_ECallbackFlagsRegistered is removed by SteamAPI_UnregisterCallback.
-			NativeMethods.SteamAPI_UnregisterCallback(m_pCCallbackBase.AddrOfPinnedObject());
+			CallbackDispatcher.Unregister(this);
+			m_bIsRegistered = false;
 		}
 
-		public void SetGameserverFlag() {
-			m_CCallbackBase.m_nCallbackFlags |= CCallbackBase.k_ECallbackFlagsGameServer;
+		public override bool IsGameServer => m_bIsRegistered;
+
+		internal override Type GetCallbackType() {
+			return typeof(T);
 		}
 
-		private void OnRunCallback(
-#if !STDCALL
-			IntPtr thisptr,
-#endif
-			IntPtr pvParam) {
+		internal override void OnRunCallback(IntPtr pvParam) {
 			try {
 				m_Func((T)Marshal.PtrToStructure(pvParam, typeof(T)));
 			}
@@ -167,61 +251,19 @@ namespace Steamworks {
 				CallbackDispatcher.ExceptionHandler(e);
 			}
 		}
+	}
 
-		// Shouldn't ever get called here, but this is what C++ Steamworks does!
-		private void OnRunCallResult(
-#if !STDCALL
-			IntPtr thisptr,
-#endif
-			IntPtr pvParam, bool bFailed, ulong hSteamAPICall) {
-			try { 
-				m_Func((T)Marshal.PtrToStructure(pvParam, typeof(T)));
-			}
-			catch (Exception e) {
-				CallbackDispatcher.ExceptionHandler(e);
-			}
-		}
-
-		private int OnGetCallbackSizeBytes(
-#if !STDCALL
-			IntPtr thisptr
-#endif
-			) {
-			return m_size;
-		}
-
-		// Steamworks.NET Specific
-		private void BuildCCallbackBase() {
-			m_CallbackBaseVTable = new CCallbackBaseVTable() {
-				m_RunCallResult = OnRunCallResult,
-				m_RunCallback = OnRunCallback,
-				m_GetCallbackSizeBytes = OnGetCallbackSizeBytes
-			};
-			m_pVTable = Marshal.AllocHGlobal(Marshal.SizeOf(typeof(CCallbackBaseVTable)));
-			Marshal.StructureToPtr(m_CallbackBaseVTable, m_pVTable, false);
-
-			m_CCallbackBase = new CCallbackBase() {
-				m_vfptr = m_pVTable,
-				m_nCallbackFlags = 0,
-				m_iCallback = CallbackIdentities.GetCallbackIdentity(typeof(T))
-			};
-			m_pCCallbackBase = GCHandle.Alloc(m_CCallbackBase, GCHandleType.Pinned);
-		}
+	public abstract class CallResult {
+		internal abstract Type GetCallbackType();
+		internal abstract void OnRunCallResult(IntPtr pvParam, bool bFailed, ulong hSteamAPICall);
 	}
 
 	public sealed class CallResult<T> : IDisposable {
-		private CCallbackBaseVTable m_CallbackBaseVTable;
-		private IntPtr m_pVTable = IntPtr.Zero;
-		private CCallbackBase m_CCallbackBase;
-		private GCHandle m_pCCallbackBase;
-
 		public delegate void APIDispatchDelegate(T param, bool bIOFailure);
 		private event APIDispatchDelegate m_Func;
 
 		private SteamAPICall_t m_hAPICall = SteamAPICall_t.Invalid;
 		public SteamAPICall_t Handle { get { return m_hAPICall; } }
-
-		private readonly int m_size = Marshal.SizeOf(typeof(T));
 
 		private bool m_bDisposed = false;
 
@@ -236,7 +278,6 @@ namespace Steamworks {
 
 		public CallResult(APIDispatchDelegate func = null) {
 			m_Func = func;
-			BuildCCallbackBase();
 		}
 
 		~CallResult() {
@@ -251,14 +292,6 @@ namespace Steamworks {
 			GC.SuppressFinalize(this);
 
 			Cancel();
-
-			if (m_pVTable != IntPtr.Zero) {
-				Marshal.FreeHGlobal(m_pVTable);
-			}
-
-			if (m_pCCallbackBase.IsAllocated) {
-				m_pCCallbackBase.Free();
-			}
 
 			m_bDisposed = true;
 		}
@@ -275,13 +308,13 @@ namespace Steamworks {
 			}
 
 			if (m_hAPICall != SteamAPICall_t.Invalid) {
-				NativeMethods.SteamAPI_UnregisterCallResult(m_pCCallbackBase.AddrOfPinnedObject(), (ulong)m_hAPICall);
+				CallbackDispatcher.Unregister(m_hAPICall, this);
 			}
 
 			m_hAPICall = hAPICall;
 
 			if (hAPICall != SteamAPICall_t.Invalid) {
-				NativeMethods.SteamAPI_RegisterCallResult(m_pCCallbackBase.AddrOfPinnedObject(), (ulong)hAPICall);
+				CallbackDispatcher.Register(hAPICall, this);
 			}
 		}
 
@@ -291,36 +324,16 @@ namespace Steamworks {
 
 		public void Cancel() {
 			if (m_hAPICall != SteamAPICall_t.Invalid) {
-				NativeMethods.SteamAPI_UnregisterCallResult(m_pCCallbackBase.AddrOfPinnedObject(), (ulong)m_hAPICall);
+				CallbackDispatcher.Unregister(m_hAPICall, this);
 				m_hAPICall = SteamAPICall_t.Invalid;
 			}
 		}
 
-		public void SetGameserverFlag() {
-			m_CCallbackBase.m_nCallbackFlags |= CCallbackBase.k_ECallbackFlagsGameServer;
+		internal override Type GetCallbackType() {
+			return typeof(T);
 		}
 
-		// Shouldn't ever get called here, but this is what C++ Steamworks does!
-		private void OnRunCallback(
-#if !STDCALL
-			IntPtr thisptr,
-#endif
-			IntPtr pvParam) {
-			m_hAPICall = SteamAPICall_t.Invalid; // Caller unregisters for us
-
-			try {
-				m_Func((T)Marshal.PtrToStructure(pvParam, typeof(T)), false);
-			}
-			catch (Exception e) {
-				CallbackDispatcher.ExceptionHandler(e);
-			}
-		}
-
-		private void OnRunCallResult(
-#if !STDCALL
-			IntPtr thisptr,
-#endif
-			IntPtr pvParam, bool bFailed, ulong hSteamAPICall_) {
+		internal override void OnRunCallResult(IntPtr pvParam, bool bFailed, ulong hSteamAPICall_) {
 			SteamAPICall_t hSteamAPICall = (SteamAPICall_t)hSteamAPICall_;
 			if (hSteamAPICall == m_hAPICall) {
 				m_hAPICall = SteamAPICall_t.Invalid; // Caller unregisters for us
@@ -333,87 +346,6 @@ namespace Steamworks {
 				}
 			}
 		}
-
-		private int OnGetCallbackSizeBytes(
-#if !STDCALL
-			IntPtr thisptr
-#endif
-			) {
-			return m_size;
-		}
-
-		// Steamworks.NET Specific
-		private void BuildCCallbackBase() {
-			m_CallbackBaseVTable = new CCallbackBaseVTable() {
-				m_RunCallback = OnRunCallback,
-				m_RunCallResult = OnRunCallResult,
-				m_GetCallbackSizeBytes = OnGetCallbackSizeBytes
-			};
-			m_pVTable = Marshal.AllocHGlobal(Marshal.SizeOf(typeof(CCallbackBaseVTable)));
-			Marshal.StructureToPtr(m_CallbackBaseVTable, m_pVTable, false);
-
-			m_CCallbackBase = new CCallbackBase() {
-				m_vfptr = m_pVTable,
-				m_nCallbackFlags = 0,
-				m_iCallback = CallbackIdentities.GetCallbackIdentity(typeof(T))
-			};
-			m_pCCallbackBase = GCHandle.Alloc(m_CCallbackBase, GCHandleType.Pinned);
-		}
-	}
-
-	[StructLayout(LayoutKind.Sequential)]
-	internal class CCallbackBase {
-		public const byte k_ECallbackFlagsRegistered = 0x01;
-		public const byte k_ECallbackFlagsGameServer = 0x02;
-
-		public IntPtr m_vfptr;
-		public byte m_nCallbackFlags;
-		public int m_iCallback;
-	}
-
-	[StructLayout(LayoutKind.Sequential)]
-	internal class CCallbackBaseVTable {
-#if STDCALL
-		private const CallingConvention cc = CallingConvention.StdCall;
-
-		[UnmanagedFunctionPointer(cc)]
-		public delegate void RunCBDel(IntPtr pvParam);
-		[UnmanagedFunctionPointer(cc)]
-		public delegate void RunCRDel(IntPtr pvParam, [MarshalAs(UnmanagedType.I1)] bool bIOFailure, ulong hSteamAPICall);
-		[UnmanagedFunctionPointer(cc)]
-		public delegate int GetCallbackSizeBytesDel();
-#else
-	#if THISCALL
-		private const CallingConvention cc = CallingConvention.ThisCall;
-	#else
-		private const CallingConvention cc = CallingConvention.Cdecl;
-	#endif
-
-		[UnmanagedFunctionPointer(cc)]
-		public delegate void RunCBDel(IntPtr thisptr, IntPtr pvParam);
-		[UnmanagedFunctionPointer(cc)]
-		public delegate void RunCRDel(IntPtr thisptr, IntPtr pvParam, [MarshalAs(UnmanagedType.I1)] bool bIOFailure, ulong hSteamAPICall);
-		[UnmanagedFunctionPointer(cc)]
-		public delegate int GetCallbackSizeBytesDel(IntPtr thisptr);
-#endif
-
-		// RunCallback and RunCallResult are swapped in MSVC ABI
-#if WINDOWS_BUILD
-		[NonSerialized]
-		[MarshalAs(UnmanagedType.FunctionPtr)]
-		public RunCRDel m_RunCallResult;
-#endif
-		[NonSerialized]
-		[MarshalAs(UnmanagedType.FunctionPtr)]
-		public RunCBDel m_RunCallback;
-#if !WINDOWS_BUILD
-		[NonSerialized]
-		[MarshalAs(UnmanagedType.FunctionPtr)]
-		public RunCRDel m_RunCallResult;
-#endif
-		[NonSerialized]
-		[MarshalAs(UnmanagedType.FunctionPtr)]
-		public GetCallbackSizeBytesDel m_GetCallbackSizeBytes;
 	}
 }
 

--- a/Plugins/Steamworks.NET/CallbackDispatcher.cs
+++ b/Plugins/Steamworks.NET/CallbackDispatcher.cs
@@ -49,19 +49,26 @@ namespace Steamworks {
 		private static object m_sync = new object();
 		private static CallbackMsg_t m_callbackMsg; // Preallocated
 		private static GCHandle m_pCallbackMsg;
+		private static int m_initCount;
 
-		public static bool IsInitialized { get; private set; }
+		public static bool IsInitialized => m_initCount > 0;
 
 		internal static void Initialize() {
-			NativeMethods.SteamAPI_ManualDispatch_Init();
-			m_pCallbackMsg = GCHandle.Alloc(m_callbackMsg, GCHandleType.Pinned);
-			IsInitialized = true;
+			lock (m_sync) {
+				if (m_initCount == 0) {
+					NativeMethods.SteamAPI_ManualDispatch_Init();
+					m_pCallbackMsg = GCHandle.Alloc(m_callbackMsg, GCHandleType.Pinned);
+				}
+				++m_initCount;
+			}
 		}
 
 		internal static void Shutdown() {
-			if (m_pCallbackMsg.IsAllocated)
-				m_pCallbackMsg.Free();
-			IsInitialized = false;
+			lock (m_sync) {
+				--m_initCount;
+				if (m_initCount == 0)
+					m_pCallbackMsg.Free();
+			}
 		}
 
 		public static void Prune() {

--- a/Plugins/Steamworks.NET/CallbackDispatcher.cs
+++ b/Plugins/Steamworks.NET/CallbackDispatcher.cs
@@ -44,8 +44,7 @@ namespace Steamworks {
 
 		private static Dictionary<int, List<Callback>> m_registeredCallbacks = new Dictionary<int, List<Callback>>();
 		private static Dictionary<int, List<Callback>> m_registeredGameServerCallbacks = new Dictionary<int, List<Callback>>();
-		private static Dictionary<ulong, CallResult> m_registeredCallResults = new Dictionary<ulong, CallResult>();
-		private static List<ulong> m_dispatchedApiCalls = new List<ulong>();
+		private static Dictionary<ulong, List<CallResult>> m_registeredCallResults = new Dictionary<ulong, List<CallResult>>();
 		private static object m_sync = new object();
 		private static CallbackMsg_t m_callbackMsg; // Preallocated
 		private static GCHandle m_pCallbackMsg;
@@ -71,31 +70,6 @@ namespace Steamworks {
 			}
 		}
 
-		public static void Prune() {
-			List<int> unusedICallbacks = new List<int>();
-			lock (m_sync) {
-				foreach (var pair in m_registeredCallbacks) {
-					if (pair.Value.Count == 0)
-						unusedICallbacks.Add(pair.Key);
-				}
-
-				foreach (var iCallback in unusedICallbacks) {
-					m_registeredCallbacks.Remove(iCallback);
-				}
-
-				unusedICallbacks.Clear();
-
-				foreach (var pair in m_registeredGameServerCallbacks) {
-					if (pair.Value.Count == 0)
-						unusedICallbacks.Add(pair.Key);
-				}
-
-				foreach (var iCallback in unusedICallbacks) {
-					m_registeredGameServerCallbacks.Remove(iCallback);
-				}
-			}
-		}
-
 		internal static void Register(Callback cb) {
 			int iCallback = CallbackIdentities.GetCallbackIdentity(cb.GetCallbackType());
 			var callbacksRegistry = cb.IsGameServer ? m_registeredGameServerCallbacks : m_registeredCallbacks;
@@ -112,9 +86,13 @@ namespace Steamworks {
 
 		internal static void Register(SteamAPICall_t asyncCall, CallResult cr) {
 			lock (m_sync) {
-				// Dispatch goes to last CallResult the SteamAPICall_t is set on
-				Unregister(asyncCall);
-				m_registeredCallResults.Add((ulong)asyncCall, cr);
+				List<CallResult> callResultsList;
+				if (!m_registeredCallResults.TryGetValue((ulong)asyncCall, out callResultsList)) {
+					callResultsList = new List<CallResult>();
+					m_registeredCallResults.Add((ulong)asyncCall, callResultsList);
+				}
+
+				callResultsList.Add(cb);
 			}
 		}
 
@@ -122,18 +100,20 @@ namespace Steamworks {
 			int iCallback = CallbackIdentities.GetCallbackIdentity(cb.GetCallbackType());
 			var callbacksRegistry = cb.IsGameServer ? m_registeredGameServerCallbacks : m_registeredCallbacks;
 			lock (m_sync) {
-				List<Callback> callbacksList;
 				if (callbacksRegistry.TryGetValue(iCallback, out var callbacksList)) {
 					callbacksList.Remove(cb);
+					if (callbacksList.Count == 0)
+						callbacksRegistry.Remove(iCallback);
 				}
 			}
 		}
 
-		internal static void Unregister(SteamAPICall_t asyncCall) {
+		internal static void Unregister(SteamAPICall_t asyncCall, CallResult cr) {
 			lock (m_sync) {
-				if (m_registeredCallResults.TryGetValue((ulong)asyncCall, out CallResult cr)) {
-					cr.SetUnregistered();
-					m_registeredCallResults.Remove((ulong)asyncCall);
+				if (m_registeredCallResults.TryGetValue((ulong)asyncCall, out var callResultsList)) {
+					callResultsList.Remove(cr);
+					if (callResultsList.Count == 0)
+						m_registeredCallResults.Remove((ulong)asyncCall);
 				}
 			}
 		}
@@ -153,12 +133,17 @@ namespace Steamworks {
 						IntPtr pTmpCallResult = Marshal.AllocHGlobal(callCompletedCb.m_cubParam);
 						bool bFailed;
 						if (NativeMethods.SteamAPI_ManualDispatch_GetAPICallResult(hSteamPipe, callCompletedCb.m_hAsyncCall, pTmpCallResult, callCompletedCb.m_cubParam, callCompletedCb.m_iCallback, out bFailed)) {
-							if (m_registeredCallResults.TryGetValue((ulong)callCompletedCb.m_hAsyncCall, out CallResult cr)) {
-								cr.OnRunCallResult(pTmpCallResult, bFailed, (ulong)callCompletedCb.m_hAsyncCall);
+							lock (m_sync) {
+								if (m_registeredCallResults.TryGetValue((ulong)callCompletedCb.m_hAsyncCall, out var callResults)) {
+									m_registeredCallResults.Remove((ulong)callCompletedCb.m_hAsyncCall);
+									foreach (var cr in callResultsCopy) {
+										cr.OnRunCallResult(pTmpCallResult, bFailed, (ulong)callCompletedCb.m_hAsyncCall);
+										cr.SetUnregistered();
+									}
+								}
 							}
 						}
 						Marshal.FreeHGlobal(pTmpCallResult);
-						m_dispatchedApiCalls.Add((ulong)callCompletedCb.m_hAsyncCall);
 					} else {
 						if (callbacksRegistry.TryGetValue(m_callbackMsg.m_iCallback, out var callbacks)) {
 							List<Callback> callbacksCopy;
@@ -174,11 +159,6 @@ namespace Steamworks {
 					ExceptionHandler(e);
 				} finally {
 					NativeMethods.SteamAPI_ManualDispatch_FreeLastCallback(hSteamPipe);
-					foreach (var call in m_dispatchedApiCalls) {
-						m_registeredCallResults[call].SetUnregistered();
-						m_registeredCallResults.Remove(call);
-					}
-					m_dispatchedApiCalls.Clear();
 				}
 			}
 		}

--- a/Plugins/Steamworks.NET/CallbackDispatcher.cs
+++ b/Plugins/Steamworks.NET/CallbackDispatcher.cs
@@ -46,8 +46,7 @@ namespace Steamworks {
 		private static Dictionary<int, List<Callback>> m_registeredGameServerCallbacks = new Dictionary<int, List<Callback>>();
 		private static Dictionary<ulong, List<CallResult>> m_registeredCallResults = new Dictionary<ulong, List<CallResult>>();
 		private static object m_sync = new object();
-		private static CallbackMsg_t m_callbackMsg; // Preallocated
-		private static GCHandle m_pCallbackMsg;
+		private static IntPtr m_pCallbackMsg;
 		private static int m_initCount;
 
 		public static bool IsInitialized => m_initCount > 0;
@@ -56,7 +55,7 @@ namespace Steamworks {
 			lock (m_sync) {
 				if (m_initCount == 0) {
 					NativeMethods.SteamAPI_ManualDispatch_Init();
-					m_pCallbackMsg = GCHandle.Alloc(m_callbackMsg, GCHandleType.Pinned);
+					m_pCallbackMsg = Marshal.AllocHGlobal(Marshal.SizeOf(typeof(CallbackMsg_t)));
 				}
 				++m_initCount;
 			}
@@ -65,8 +64,11 @@ namespace Steamworks {
 		internal static void Shutdown() {
 			lock (m_sync) {
 				--m_initCount;
-				if (m_initCount == 0)
-					m_pCallbackMsg.Free();
+				if (m_initCount == 0) {
+					UnregisterAll();
+					Marshal.FreeHGlobal(m_pCallbackMsg);
+					m_pCallbackMsg = IntPtr.Zero;
+				}
 			}
 		}
 
@@ -92,7 +94,7 @@ namespace Steamworks {
 					m_registeredCallResults.Add((ulong)asyncCall, callResultsList);
 				}
 
-				callResultsList.Add(cb);
+				callResultsList.Add(cr);
 			}
 		}
 
@@ -118,25 +120,54 @@ namespace Steamworks {
 			}
 		}
 
+		private static void UnregisterAll() {
+			List<Callback> callbacks = new List<Callback>();
+			List<CallResult> callResults = new List<CallResult>();
+			lock (m_sync) {
+				foreach (var pair in m_registeredCallbacks) {
+					callbacks.AddRange(pair.Value);
+				}
+				m_registeredCallbacks.Clear();
+
+				foreach (var pair in m_registeredGameServerCallbacks) {
+					callbacks.AddRange(pair.Value);
+				}
+				m_registeredGameServerCallbacks.Clear();
+
+				foreach (var pair in m_registeredCallResults) {
+					callResults.AddRange(pair.Value);
+				}
+				m_registeredCallResults.Clear();
+
+				foreach (var callback in callbacks) {
+					callback.SetUnregistered();
+				}
+
+				foreach (var callResult in callResults) {
+					callResult.SetUnregistered();
+				}
+			}
+		}
+
 		internal static void RunFrame(bool isGameServer) {
 			if (!IsInitialized) throw new InvalidOperationException("Callback dispatcher is not initialized.");
 
 			HSteamPipe hSteamPipe = (HSteamPipe)(isGameServer ? NativeMethods.SteamGameServer_GetHSteamPipe() : NativeMethods.SteamAPI_GetHSteamPipe());
 			NativeMethods.SteamAPI_ManualDispatch_RunFrame(hSteamPipe);
 			var callbacksRegistry = isGameServer ? m_registeredGameServerCallbacks : m_registeredCallbacks;
-			while (NativeMethods.SteamAPI_ManualDispatch_GetNextCallback(hSteamPipe, m_pCallbackMsg.AddrOfPinnedObject())) {
+			while (NativeMethods.SteamAPI_ManualDispatch_GetNextCallback(hSteamPipe, m_pCallbackMsg)) {
+				CallbackMsg_t callbackMsg = (CallbackMsg_t)Marshal.PtrToStructure(m_pCallbackMsg, typeof(CallbackMsg_t));
 				try {
 					// Check for dispatching API call results
-					if (m_callbackMsg.m_iCallback == SteamAPICallCompleted_t.k_iCallback) {
-						SteamAPICallCompleted_t callCompletedCb;
-						Marshal.PtrToStructure(m_callbackMsg.m_pubParam, callCompletedCb);
-						IntPtr pTmpCallResult = Marshal.AllocHGlobal(callCompletedCb.m_cubParam);
+					if (callbackMsg.m_iCallback == SteamAPICallCompleted_t.k_iCallback) {
+						SteamAPICallCompleted_t callCompletedCb = (SteamAPICallCompleted_t)Marshal.PtrToStructure(callbackMsg.m_pubParam, typeof(SteamAPICallCompleted_t));
+						IntPtr pTmpCallResult = Marshal.AllocHGlobal((int)callCompletedCb.m_cubParam);
 						bool bFailed;
-						if (NativeMethods.SteamAPI_ManualDispatch_GetAPICallResult(hSteamPipe, callCompletedCb.m_hAsyncCall, pTmpCallResult, callCompletedCb.m_cubParam, callCompletedCb.m_iCallback, out bFailed)) {
+						if (NativeMethods.SteamAPI_ManualDispatch_GetAPICallResult(hSteamPipe, callCompletedCb.m_hAsyncCall, pTmpCallResult, (int)callCompletedCb.m_cubParam, callCompletedCb.m_iCallback, out bFailed)) {
 							lock (m_sync) {
 								if (m_registeredCallResults.TryGetValue((ulong)callCompletedCb.m_hAsyncCall, out var callResults)) {
 									m_registeredCallResults.Remove((ulong)callCompletedCb.m_hAsyncCall);
-									foreach (var cr in callResultsCopy) {
+									foreach (var cr in callResults) {
 										cr.OnRunCallResult(pTmpCallResult, bFailed, (ulong)callCompletedCb.m_hAsyncCall);
 										cr.SetUnregistered();
 									}
@@ -145,13 +176,13 @@ namespace Steamworks {
 						}
 						Marshal.FreeHGlobal(pTmpCallResult);
 					} else {
-						if (callbacksRegistry.TryGetValue(m_callbackMsg.m_iCallback, out var callbacks)) {
+						if (callbacksRegistry.TryGetValue(callbackMsg.m_iCallback, out var callbacks)) {
 							List<Callback> callbacksCopy;
 							lock (m_sync) {
 								callbacksCopy = new List<Callback>(callbacks);
 							}
 							foreach (var callback in callbacksCopy) {
-								callback.OnRunCallback(m_callbackMsg.m_pubParam);
+								callback.OnRunCallback(callbackMsg.m_pubParam);
 							}
 						}
 					}
@@ -168,6 +199,7 @@ namespace Steamworks {
 		public abstract bool IsGameServer { get; }
 		internal abstract Type GetCallbackType();
 		internal abstract void OnRunCallback(IntPtr pvParam);
+		internal abstract void SetUnregistered();
 	}
 
 	public sealed class Callback<T> : Callback, IDisposable {
@@ -213,7 +245,8 @@ namespace Steamworks {
 
 			GC.SuppressFinalize(this);
 
-			Unregister();
+			if (m_bIsRegistered)
+				Unregister();
 
 			m_bDisposed = true;
 		}
@@ -253,6 +286,10 @@ namespace Steamworks {
 				CallbackDispatcher.ExceptionHandler(e);
 			}
 		}
+
+		internal override void SetUnregistered() {
+			m_bIsRegistered = false;
+		}
 	}
 
 	public abstract class CallResult {
@@ -261,7 +298,7 @@ namespace Steamworks {
 		internal abstract void SetUnregistered();
 	}
 
-	public sealed class CallResult<T> : IDisposable {
+	public sealed class CallResult<T> : CallResult, IDisposable {
 		public delegate void APIDispatchDelegate(T param, bool bIOFailure);
 		private event APIDispatchDelegate m_Func;
 

--- a/Plugins/Steamworks.NET/CallbackDispatcher.cs
+++ b/Plugins/Steamworks.NET/CallbackDispatcher.cs
@@ -175,6 +175,7 @@ namespace Steamworks {
 				} finally {
 					NativeMethods.SteamAPI_ManualDispatch_FreeLastCallback(hSteamPipe);
 					foreach (var call in m_dispatchedApiCalls) {
+						m_registeredCallResults[call].SetUnregistered();
 						m_registeredCallResults.Remove(call);
 					}
 					m_dispatchedApiCalls.Clear();
@@ -345,10 +346,8 @@ namespace Steamworks {
 		}
 
 		public void Cancel() {
-			if (m_hAPICall != SteamAPICall_t.Invalid) {
+			if (IsActive())
 				CallbackDispatcher.Unregister(m_hAPICall, this);
-				m_hAPICall = SteamAPICall_t.Invalid;
-			}
 		}
 
 		internal override Type GetCallbackType() {
@@ -358,8 +357,6 @@ namespace Steamworks {
 		internal override void OnRunCallResult(IntPtr pvParam, bool bFailed, ulong hSteamAPICall_) {
 			SteamAPICall_t hSteamAPICall = (SteamAPICall_t)hSteamAPICall_;
 			if (hSteamAPICall == m_hAPICall) {
-				m_hAPICall = SteamAPICall_t.Invalid; // Caller unregisters for us
-
 				try {
 					m_Func((T)Marshal.PtrToStructure(pvParam, typeof(T)), bFailed);
 				}

--- a/Plugins/Steamworks.NET/CallbackDispatcher.cs
+++ b/Plugins/Steamworks.NET/CallbackDispatcher.cs
@@ -122,41 +122,43 @@ namespace Steamworks {
 			HSteamPipe hSteamPipe = (HSteamPipe)(isGameServer ? NativeMethods.SteamGameServer_GetHSteamPipe() : NativeMethods.SteamAPI_GetHSteamPipe());
 			NativeMethods.SteamAPI_ManualDispatch_RunFrame(hSteamPipe);
 			var callbacksRegistry = isGameServer ? m_registeredGameServerCallbacks : m_registeredCallbacks;
-			lock (m_sync) {
-				while (NativeMethods.SteamAPI_ManualDispatch_GetNextCallback(hSteamPipe, m_pCallbackMsg.AddrOfPinnedObject())) {
-					try {
-						// Check for dispatching API call results
-						if (m_callbackMsg.m_iCallback == SteamAPICallCompleted_t.k_iCallback) {
-							SteamAPICallCompleted_t callCompletedCb;
-							Marshal.PtrToStructure(m_callbackMsg.m_pubParam, callCompletedCb);
-							IntPtr pTmpCallResult = Marshal.AllocHGlobal(callCompletedCb.m_cubParam);
-							bool bFailed;
-							try {
-								if (NativeMethods.SteamAPI_ManualDispatch_GetAPICallResult(hSteamPipe, callCompletedCb.m_hAsyncCall, pTmpCallResult, callCompletedCb.m_cubParam, callCompletedCb.m_iCallback, out bFailed)) {
-									if (m_registeredCallResults.TryGetValue((ulong)callCompletedCb.m_hAsyncCall, out CallResult cr)) {
-										cr.OnRunCallResult(pTmpCallResult, bFailed, (ulong)callCompletedCb.m_hAsyncCall);
-									}
-								}
-							} finally {
-								Marshal.FreeHGlobal(pTmpCallResult);
-								m_dispatchedApiCalls.Add((ulong)callCompletedCb.m_hAsyncCall);
-							}
-						} else {
-							if (callbacksRegistry.TryGetValue(m_callbackMsg.m_iCallback, out var callbacks)) {
-								foreach (var callback in callbacks) {
-									callback.OnRunCallback(m_callbackMsg.m_pubParam);
+			while (NativeMethods.SteamAPI_ManualDispatch_GetNextCallback(hSteamPipe, m_pCallbackMsg.AddrOfPinnedObject())) {
+				try {
+					// Check for dispatching API call results
+					if (m_callbackMsg.m_iCallback == SteamAPICallCompleted_t.k_iCallback) {
+						SteamAPICallCompleted_t callCompletedCb;
+						Marshal.PtrToStructure(m_callbackMsg.m_pubParam, callCompletedCb);
+						IntPtr pTmpCallResult = Marshal.AllocHGlobal(callCompletedCb.m_cubParam);
+						bool bFailed;
+						try {
+							if (NativeMethods.SteamAPI_ManualDispatch_GetAPICallResult(hSteamPipe, callCompletedCb.m_hAsyncCall, pTmpCallResult, callCompletedCb.m_cubParam, callCompletedCb.m_iCallback, out bFailed)) {
+								if (m_registeredCallResults.TryGetValue((ulong)callCompletedCb.m_hAsyncCall, out CallResult cr)) {
+									cr.OnRunCallResult(pTmpCallResult, bFailed, (ulong)callCompletedCb.m_hAsyncCall);
 								}
 							}
+						} finally {
+							Marshal.FreeHGlobal(pTmpCallResult);
+							m_dispatchedApiCalls.Add((ulong)callCompletedCb.m_hAsyncCall);
 						}
-					} catch (Exception e) {
-						ExceptionHandler(e);
-					} finally {
-						NativeMethods.SteamAPI_ManualDispatch_FreeLastCallback(hSteamPipe);
-						foreach (var call in m_dispatchedApiCalls) {
-							m_registeredCallResults.Remove(call);
+					} else {
+						if (callbacksRegistry.TryGetValue(m_callbackMsg.m_iCallback, out var callbacks)) {
+							List<Callback> callbacksCopy;
+							lock (m_sync) {
+								callbacksCopy = new List<Callback>(callbacks);
+							}
+							foreach (var callback in callbacksCopy) {
+								callback.OnRunCallback(m_callbackMsg.m_pubParam);
+							}
 						}
-						m_dispatchedApiCalls.Clear();
 					}
+				} catch (Exception e) {
+					ExceptionHandler(e);
+				} finally {
+					NativeMethods.SteamAPI_ManualDispatch_FreeLastCallback(hSteamPipe);
+					foreach (var call in m_dispatchedApiCalls) {
+						m_registeredCallResults.Remove(call);
+					}
+					m_dispatchedApiCalls.Clear();
 				}
 			}
 		}

--- a/Plugins/Steamworks.NET/CallbackDispatcher.cs
+++ b/Plugins/Steamworks.NET/CallbackDispatcher.cs
@@ -106,7 +106,8 @@ namespace Steamworks {
 		internal static void Register(SteamAPICall_t asyncCall, CallResult cr) {
 			lock (m_sync) {
 				// Dispatch goes to last CallResult the SteamAPICall_t is set on
-				m_registeredCallResults[(ulong)asyncCall] = cr;
+				Unregister(asyncCall);
+				m_registeredCallResults.Add((ulong)asyncCall, cr);
 			}
 		}
 
@@ -123,7 +124,10 @@ namespace Steamworks {
 
 		internal static void Unregister(SteamAPICall_t asyncCall) {
 			lock (m_sync) {
-				m_registeredCallResults.Remove((ulong)asyncCall);
+				if (m_registeredCallResults.TryGetValue((ulong)asyncCall, out CallResult cr)) {
+					cr.SetUnregistered();
+					m_registeredCallResults.Remove((ulong)asyncCall);
+				}
 			}
 		}
 
@@ -250,7 +254,7 @@ namespace Steamworks {
 			m_bIsRegistered = false;
 		}
 
-		public override bool IsGameServer => m_bIsRegistered;
+		public override bool IsGameServer => m_bGameServer;
 
 		internal override Type GetCallbackType() {
 			return typeof(T);
@@ -269,6 +273,7 @@ namespace Steamworks {
 	public abstract class CallResult {
 		internal abstract Type GetCallbackType();
 		internal abstract void OnRunCallResult(IntPtr pvParam, bool bFailed, ulong hSteamAPICall);
+		internal abstract void SetUnregistered();
 	}
 
 	public sealed class CallResult<T> : IDisposable {
@@ -358,6 +363,10 @@ namespace Steamworks {
 					CallbackDispatcher.ExceptionHandler(e);
 				}
 			}
+		}
+
+		internal override void SetUnregistered() {
+			m_hAPICall = SteamAPICall_t.Invalid;
 		}
 	}
 }

--- a/Plugins/Steamworks.NET/CallbackDispatcher.cs
+++ b/Plugins/Steamworks.NET/CallbackDispatcher.cs
@@ -145,16 +145,13 @@ namespace Steamworks {
 						Marshal.PtrToStructure(m_callbackMsg.m_pubParam, callCompletedCb);
 						IntPtr pTmpCallResult = Marshal.AllocHGlobal(callCompletedCb.m_cubParam);
 						bool bFailed;
-						try {
-							if (NativeMethods.SteamAPI_ManualDispatch_GetAPICallResult(hSteamPipe, callCompletedCb.m_hAsyncCall, pTmpCallResult, callCompletedCb.m_cubParam, callCompletedCb.m_iCallback, out bFailed)) {
-								if (m_registeredCallResults.TryGetValue((ulong)callCompletedCb.m_hAsyncCall, out CallResult cr)) {
-									cr.OnRunCallResult(pTmpCallResult, bFailed, (ulong)callCompletedCb.m_hAsyncCall);
-								}
+						if (NativeMethods.SteamAPI_ManualDispatch_GetAPICallResult(hSteamPipe, callCompletedCb.m_hAsyncCall, pTmpCallResult, callCompletedCb.m_cubParam, callCompletedCb.m_iCallback, out bFailed)) {
+							if (m_registeredCallResults.TryGetValue((ulong)callCompletedCb.m_hAsyncCall, out CallResult cr)) {
+								cr.OnRunCallResult(pTmpCallResult, bFailed, (ulong)callCompletedCb.m_hAsyncCall);
 							}
-						} finally {
-							Marshal.FreeHGlobal(pTmpCallResult);
-							m_dispatchedApiCalls.Add((ulong)callCompletedCb.m_hAsyncCall);
 						}
+						Marshal.FreeHGlobal(pTmpCallResult);
+						m_dispatchedApiCalls.Add((ulong)callCompletedCb.m_hAsyncCall);
 					} else {
 						if (callbacksRegistry.TryGetValue(m_callbackMsg.m_iCallback, out var callbacks)) {
 							List<Callback> callbacksCopy;

--- a/Plugins/Steamworks.NET/Steam.cs
+++ b/Plugins/Steamworks.NET/Steam.cs
@@ -47,6 +47,9 @@ namespace Steamworks {
 				ret = CSteamAPIContext.Init();
 			}
 
+			if (ret)
+				CallbackDispatcher.Initialize();
+
 			return ret;
 		}
 
@@ -55,6 +58,7 @@ namespace Steamworks {
 			InteropHelp.TestIfPlatformSupported();
 			NativeMethods.SteamAPI_Shutdown();
 			CSteamAPIContext.Clear();
+			CallbackDispatcher.Shutdown();
 		}
 
 		// SteamAPI_RestartAppIfNecessary ensures that your executable was launched through Steam.
@@ -113,8 +117,7 @@ namespace Steamworks {
 		// One alternative is to call SteamAPI_RunCallbacks from the main thread only,
 		// and call SteamAPI_ReleaseCurrentThreadMemory regularly on other threads.
 		public static void RunCallbacks() {
-			InteropHelp.TestIfPlatformSupported();
-			NativeMethods.SteamAPI_RunCallbacks();
+			CallbackDispatcher.RunFrame(false);
 		}
 
 		//----------------------------------------------------------------------------------------------------------------------------------------------------------//
@@ -170,6 +173,9 @@ namespace Steamworks {
 				ret = CSteamGameServerAPIContext.Init();
 			}
 
+			if (ret)
+				CallbackDispatcher.Initialize();
+
 			return ret;
 		}
 
@@ -177,11 +183,11 @@ namespace Steamworks {
 			InteropHelp.TestIfPlatformSupported();
 			NativeMethods.SteamGameServer_Shutdown();
 			CSteamGameServerAPIContext.Clear();
+			CallbackDispatcher.Shutdown();
 		}
 
 		public static void RunCallbacks() {
-			InteropHelp.TestIfPlatformSupported();
-			NativeMethods.SteamGameServer_RunCallbacks();
+			CallbackDispatcher.RunFrame(true);
 		}
 
 		// Most Steam API functions allocate some amount of thread-local memory for

--- a/Plugins/Steamworks.NET/autogen/NativeMethods.cs
+++ b/Plugins/Steamworks.NET/autogen/NativeMethods.cs
@@ -121,6 +121,10 @@ namespace Steamworks {
 
 		[DllImport(NativeLibraryName, EntryPoint = "SteamAPI_ManualDispatch_FreeLastCallback", CallingConvention = CallingConvention.Cdecl)]
 		public static extern void SteamAPI_ManualDispatch_FreeLastCallback(HSteamPipe hSteamPipe);
+
+		[DllImport(NativeLibraryName, EntryPoin = "SteamAPI_ManualDispatch_GetAPICallResult", CallingConvention = CallingConvention.Cdecl)]
+		[return: MarshalAs(UnmanagedType.I1)]
+		public static extern bool SteamAPI_ManualDispatch_GetAPICallResult(HSteamPipe hSteamPipe, SteamAPICall_t hSteamAPICall, IntPtr pCallback, int cubCallback, int iCallbackExpected, out bool pbFailed);
 #endregion
 #region steam_gameserver.h
 		[DllImport(NativeLibraryName, EntryPoint = "SteamGameServer_InitSafe", CallingConvention = CallingConvention.Cdecl)]

--- a/Plugins/Steamworks.NET/autogen/NativeMethods.cs
+++ b/Plugins/Steamworks.NET/autogen/NativeMethods.cs
@@ -122,7 +122,7 @@ namespace Steamworks {
 		[DllImport(NativeLibraryName, EntryPoint = "SteamAPI_ManualDispatch_FreeLastCallback", CallingConvention = CallingConvention.Cdecl)]
 		public static extern void SteamAPI_ManualDispatch_FreeLastCallback(HSteamPipe hSteamPipe);
 
-		[DllImport(NativeLibraryName, EntryPoin = "SteamAPI_ManualDispatch_GetAPICallResult", CallingConvention = CallingConvention.Cdecl)]
+		[DllImport(NativeLibraryName, EntryPoint = "SteamAPI_ManualDispatch_GetAPICallResult", CallingConvention = CallingConvention.Cdecl)]
 		[return: MarshalAs(UnmanagedType.I1)]
 		public static extern bool SteamAPI_ManualDispatch_GetAPICallResult(HSteamPipe hSteamPipe, SteamAPICall_t hSteamAPICall, IntPtr pCallback, int cubCallback, int iCallbackExpected, out bool pbFailed);
 #endregion

